skimage/__init__.py

```diff
@@ -73,7 +16,33 @@
 
 
 def __dir__():
-    return __lazy_dir__() + ['__version__']
+    """Add lazy-loaded attributes to keep consistent with `__getattr__`."""
+    patched_dir = {*globals().keys(), *__lazy_dir__()}
+    return sorted(patched_dir)
```

Comment:
What is fixed by this change?

Reply:
Lazy loading expands the module attributes that are available via `__getattr__`. These new attributes are missing from the default `__dir__()` / `dir(skimage)`. If we use `__dir__` returned from  `attach_stub`, then a whole bunch of pre-existing attributes like `__file__`, `__doc__`, etc. are missing. Both are inconsistent and it seemed like a good opportunity to preempt weird or confusing behavior by code inspection tools, etc. that don't expect messing with `__dir__`. 

Spelling this out, I realize that this is also a problem with our subpackages that just use `__dir__` from `attach_stub` as is. Not sure if it's worth it to go through all these. Neither am I sure how to conveniently address this from lazy_loader's site. `globals()` returns the scope from where the function is defined, so lazy_loader would need a different way to access the correct global scope...

Reply:
Why would we want to expose things like `__file__` etc.? `__dir__` looks exactly like what I would expect atm:

```
In [1]: dir(ski)
Out[1]:
['__version__',
 'color',
 'data',
 'draw',
 'exposure',
 'feature',
 'filters',
 'future',
 'graph',
 'io',
 'measure',
 'metrics',
 'morphology',
 'registration',
 'restoration',
 'segmentation',
 'transform',
 'util']
```

You can use various hacks to get hold of other attributes, but these are the ones we want our users to see.

Reply:
If you really need to expose more (and I don't think we should), then you could:

```
def __dir__():
    return [some stuff] + __lazy_dir__()
```

Reply:
> Why would we want to expose things like __file__ etc.?

For the reasons stated above. Hiding builtin attributes feels very magical and unexpected to me, and like something that might lead confusing bugs. I'd like tab-completion in ipython to show me everything that's available and not a curated list.

Everyone is aware that `_` means private and is used to `__builtin_stuff__`. Why would we need to hide these?

Reply:
Well, that's the beauty of `__dir__`: you can show users only the things they need to see. Advanced users can still get hold of the weird attributes they want, they're still there.

Reply:
(See `dir(scipy)` as well.)

Reply:
I see that this question was also raised in https://github.com/scipy/scipy/pull/15230#issuecomment-1006026070 where you suggested `scipy.__dict__.keys()` as a workaround. That doesn't include lazily loaded stuff and it this solution wasn't immediately obvious to me. 

I feel like I do recognize all your arguments, I just don't find them as convincing as you do. :wink: Patching `__dir__` still feels like a definsive way to preemt confusion and fits with my mental model. But since this starts to feel like bikeshedding I'm just gonna say I'm +0.5 on patching `__dir__`. If you or someone else feels more strongly in the other direction, I'm happy to resolve this either way.

Reply:
> I feel like I do recognize all your arguments, I just don't find them as convincing as you do.

This is a common problem I run into 🙃

Since you'd like to see a change in behavior, I think you should make a convincing case for (a) why this is useful (and not confusing) to the user and (b) that it does not introduce unnecessary complexity (we could improve the code to make that the case, so I'm not too worried about this part).

Reply:
Maybe more importantly, we are way overdue on the release, so a friendly request that we merge this PR without the change to `__dir__` and then do additional follow-up once we come to an agreement.

Reply:
I've added it as a talking point to today's meeting notes. 

Anyway, I'm happy to iterate on this particular issue in another PR. And I'm not seeing this issue as a blocker for the release in the first place.

Reply:
Reverted to status quo in https://github.com/scikit-image/scikit-image/pull/6892/commits/0a275cdf3e46162474f0bf0dd73819894f38b17d with regards to hiding stuff in `__dir__`.

Reply:
[Python's docs](https://docs.python.org/3/library/functions.html#dir) seem to be compatible with your stance @stefanv:

> The resulting list is not necessarily complete and may be inaccurate when the object has a custom `__getattr__()`.

> [...] it attempts to produce the most relevant, rather than complete, information: [...]

> Because `dir()` is supplied primarily as a convenience for use at an interactive prompt, it tries to supply an interesting set of names more than it tries to supply a rigorously or consistently defined set of names, and its detailed behavior may change across releases.